dataplane/handlers/interface.go

```diff
@@ -0,0 +1,293 @@
+	log.Info("starting interface handler")
+	b := &ocpath.Batch{}
+
+	if err := ni.setupPorts(ctx); err != nil {
+		return fmt.Errorf("failed to setup ports: %v", err)
+	}
+
+	b.AddPaths(
+		ocpath.Root().InterfaceAny().Enabled().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Name().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Ethernet().MacAddress().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv4().AddressAny().Ip().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv4().AddressAny().PrefixLength().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv6().AddressAny().Ip().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv6().AddressAny().PrefixLength().Config().PathStruct(),
+	)
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	ni.watchCancelFn = cancelFn
+
+	watcher := ygnmi.Watch(cancelCtx, ni.c, b.Config(), func(val *ygnmi.Value[*oc.Root]) error {
+		log.Info("reconciling interfaces")
+		root, ok := val.Val()
+		if !ok || root.Interface == nil {
+			return ygnmi.Continue
+		}
+		for _, i := range root.Interface {
+			ni.reconcile(i)
+		}
+		return ygnmi.Continue
+	})
+
+	ni.linkDoneCh = make(chan struct{})
+	linkUpdateCh := make(chan netlink.LinkUpdate)
+	ni.addrDoneCh = make(chan struct{})
+	addrUpdateCh := make(chan netlink.AddrUpdate)
+
+	if err := netlink.LinkSubscribe(linkUpdateCh, ni.linkDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to link: %v", err)
+	}
+	if err := netlink.AddrSubscribe(addrUpdateCh, ni.addrDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to addr: %v", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case up := <-linkUpdateCh:
+				ni.handleLinkUpdate(&up)
+			case up := <-addrUpdateCh:
+				ni.handleAddrUpdate(&up)
+			}
+		}
+	}()
+
+	go func() {
+		// TODO: handle error
+		if _, err := watcher.Await(); err != nil {
+			log.Warningf("interface watch err: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops all watchers.
+func (ni *Interface) Stop() {
+	ni.watchCancelFn()
+	close(ni.linkDoneCh)
+	close(ni.addrDoneCh)
+}
+
+// reconcile compares the interface config with state and modifies state to match config.
+func (ni *Interface) reconcile(config *oc.Interface) {
+	ni.mu.RLock()
+	defer ni.mu.RUnlock()
+
+	tapName := engine.IntfNameToTapName(config.GetName())
+	state := ni.state[config.GetName()]
+
+	if config.GetOrCreateEthernet().MacAddress != nil {
+		if config.GetEthernet().GetMacAddress() != state.GetEthernet().GetMacAddress() {
+			log.Infof("setting interface %s hw-addr %q", engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress())
+			if err := kernel.SetInterfaceHWAddr(engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress()); err != nil {
+				log.Warningf("Failed to set mac address of port: %v", err)
+				return
+			}
+		}
+	}
+	if config.GetOrCreateSubinterface(0).Enabled != nil {
```

Comment:
What if mac-address is nil? Should the TAP interface be deleted?

Reply:
why would remove an interface?